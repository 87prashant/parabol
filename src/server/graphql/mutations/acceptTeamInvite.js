import {GraphQLID, GraphQLNonNull} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import parseInviteToken from 'server/graphql/mutations/helpers/inviteTeamMembers/parseInviteToken';
import validateInviteTokenKey from 'server/graphql/mutations/helpers/inviteTeamMembers/validateInviteTokenKey';
import AcceptTeamInvitePayload from 'server/graphql/types/AcceptTeamInvitePayload';
import acceptTeamInvite from 'server/safeMutations/acceptTeamInvite';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {getUserId, isAuthenticated} from 'server/utils/authorization';
import publish from 'server/utils/publish';
import tmsSignToken from 'server/utils/tmsSignToken';
import {INVITATION, NEW_AUTH_TOKEN, TASK, TEAM, TEAM_MEMBER, UPDATED} from 'universal/utils/constants';
import toTeamMemberId from 'universal/utils/relay/toTeamMemberId';
import getActiveTeamMembersByTeamIds from 'server/safeQueries/getActiveTeamMembersByTeamIds';
import {
  sendInvitationExpiredError,
  sendInvitationHashFailError,
  sendNotAuthenticatedAccessError,
  sendTeamAlreadyJoinedError
} from 'server/utils/authorizationErrors';
import {
  sendInvitationNotFoundError, sendNoInvitationProvidedError,
  sendNotificationAccessError
} from 'server/utils/docNotFoundErrors';


const validateInviteToken = async (inviteToken, authToken) => {
  const r = getRethink();
  const now = new Date();
  const {id: inviteId, key: tokenKey} = parseInviteToken(inviteToken);

  // see if the invitation exists
  const invitation = await r.table('Invitation').get(inviteId).update({
    tokenExpiration: new Date(0),
    updatedAt: now
  }, {returnChanges: true})('changes')(0)('old_val').default(null);

  if (!invitation) {
    return sendInvitationNotFoundError(authToken, inviteToken);
  }

  const {tokenExpiration, hashedToken, teamId} = invitation;
  // see if the invitation has expired
  if (tokenExpiration < now) {
    return sendInvitationExpiredError(authToken, inviteToken);
  }

  // see if the invitation hash is valid
  const isCorrectToken = await validateInviteTokenKey(tokenKey, hashedToken);
  if (!isCorrectToken) {
    return sendInvitationHashFailError(authToken, inviteToken);
  }
  const oldtms = authToken.tms || [];
  // Check if TeamMember already exists (i.e. user invited themselves):
  const teamMemberExists = oldtms.includes(teamId);
  if (teamMemberExists) {
    return sendTeamAlreadyJoinedError(authToken, inviteToken);
  }
  return invitation;
};

export default {
  type: new GraphQLNonNull(AcceptTeamInvitePayload),
  description: `Add a user to a Team given an invitationToken or the notification id of the invitation.
    If the invitationToken is valid, returns the auth token with the new team added to tms.
    Side effect: deletes all other outstanding invitations for user.`,
  args: {
    inviteToken: {
      type: GraphQLID,
      description: 'The invitation token (first 6 bytes are the id, next 8 are the pre-hash)'
    },
    notificationId: {
      type: GraphQLID,
      description: 'The notification id of the team invite'
    }
  },
  async resolve(source, {inviteToken, notificationId}, {authToken, dataLoader, socketId: mutatorId}) {
    const r = getRethink();
    const operationId = dataLoader.share();
    const subOptions = {mutatorId, operationId};

    // AUTH
    const viewerId = getUserId(authToken);
    if (!isAuthenticated(authToken)) return sendNotAuthenticatedAccessError();

    // VALIDATION
    let inviteeEmail;
    let teamId;
    if (inviteToken) {
      const errorOrInvitation = await validateInviteToken(inviteToken);
      if (errorOrInvitation.error) return errorOrInvitation;
      inviteeEmail = errorOrInvitation.email;
      teamId = errorOrInvitation.teamId;
    } else if (notificationId) {
      const notification = await r.table('Notification').get(notificationId);
      if (!notification || !notification.userIds.includes(viewerId)) return sendNotificationAccessError(authToken, notificationId);
      inviteeEmail = notification.inviteeEmail;
      teamId = notification.teamId;
    } else {
      return sendNoInvitationProvidedError(authToken);
    }

    // RESOLUTION
    const {
      hardenedTasks,
      removedNotification,
      removedInvitationId: invitationId,
      removedSoftTeamMember
    } = await acceptTeamInvite(teamId, authToken, inviteeEmail);
    const oldTMS = authToken.tms || [];
    const tms = oldTMS.concat(teamId);
    const teamMemberId = toTeamMemberId(teamId, viewerId);
    const newAuthToken = tmsSignToken(authToken, tms);

    const data = {
      userId: viewerId,
      teamId,
      teamMemberId,
      removedNotification,
      invitationId,
      softTeamMemberId: removedSoftTeamMember.id,
      taskIds: hardenedTasks.map(({id}) => id)
    };

    if (hardenedTasks.length > 0) {
      const teamMembers = await getActiveTeamMembersByTeamIds(teamId, dataLoader);
      teamMembers.forEach(({userId}) => {
        publish(TASK, userId, AcceptTeamInvitePayload, data, subOptions);
      });
    }
    // Send the new team member a welcome & a new token
    publish(NEW_AUTH_TOKEN, viewerId, UPDATED, {tms});
    auth0ManagementClient.users.updateAppMetadata({id: viewerId}, {tms});

    // remove the old invitation
    publish(INVITATION, teamId, AcceptTeamInvitePayload, data, subOptions);

    // Tell the new team member about the team, welcome them, and remove their outstanding invitation notifications
    publish(TEAM, viewerId, AcceptTeamInvitePayload, data, subOptions);

    // Tell the rest of the team about the new team member, toast the event, and remove their old invitations
    publish(TEAM_MEMBER, teamId, AcceptTeamInvitePayload, data, subOptions);

    return {...data, authToken: newAuthToken};
  }
};

