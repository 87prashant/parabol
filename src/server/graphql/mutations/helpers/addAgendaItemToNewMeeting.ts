import {AGENDA_ITEMS} from 'universal/utils/constants'
import getRethink from 'server/database/rethinkDriver'
import {DataLoaderWorker} from 'server/graphql/graphql'
import Meeting from 'server/database/types/Meeting'
import AgendaItemsPhase from 'server/database/types/AgendaItemsPhase'
import AgendaItemsStage from 'server/database/types/AgendaItemsStage'

/*
 * NewMeetings have a predefined set of stages, we need to add the new agenda item manually
 */
const addAgendaItemToNewMeeting = async (
  agendaItemId,
  teamId: string,
  dataLoader: DataLoaderWorker
) => {
  const now = new Date()
  const r = getRethink()
  const team = await dataLoader.get('teams').load(teamId)
  const {meetingId} = team
  if (!meetingId) return undefined
  // make sure it's a new meeting
  const newMeeting = (await r
    .table('NewMeeting')
    .get(meetingId)
    .default(null)) as Meeting
  if (!newMeeting) return undefined
  const {phases} = newMeeting
  const agendaItemPhase = phases.find((phase) => phase.phaseType === AGENDA_ITEMS) as
    | AgendaItemsPhase
    | undefined
  if (!agendaItemPhase) return undefined

  const {stages} = agendaItemPhase
  const newStage = new AgendaItemsStage(agendaItemId)
  newStage.isNavigable = true
  newStage.isNavigableByFacilitator = true
  stages.push(newStage)
  await r
    .table('NewMeeting')
    .get(meetingId)
    .update({
      phases,
      updatedAt: now
    })
  return meetingId
}

export default addAgendaItemToNewMeeting
