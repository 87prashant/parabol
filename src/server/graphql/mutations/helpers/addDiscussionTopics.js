import getRethink from 'server/database/rethinkDriver';
import shortid from 'shortid';
import {DISCUSS, TEAM} from 'universal/utils/constants';
import publish from 'server/utils/publish';
import StartNewMeetingPayload from 'server/graphql/types/StartNewMeetingPayload';


export const makeDiscussionStage = (reflectionGroupId, meetingId) => ({
  id: shortid.generate(),
  meetingId,
  isComplete: false,
  phaseType: DISCUSS,
  reflectionGroupId
});

const mapGroupsToStages = (reflectionGroups) => {
  const importantReflectionGroups = reflectionGroups.filter((group) => group.voterIds.length > 0);
  // handle edge case that no one votes
  if (importantReflectionGroups.length === 0) return reflectionGroups;
  importantReflectionGroups.sort((a, b) => a.voterIds.length < b.voterIds.length ? 1 : -1);
  return importantReflectionGroups;
};

const addDiscussionTopics = async (meeting, dataLoader, subOptions) => {
  const r = getRethink();
  const now = new Date();
  const {phases, teamId} = meeting;
  const discussPhase = phases.find((phase) => phase.phaseType === DISCUSS);
  if (!discussPhase) return false;
  const {id: meetingId} = meeting;
  const reflectionGroups = await dataLoader.get('retroReflectionGroupsByMeetingId').load(meetingId);
  const importantReflectionGroups = mapGroupsToStages(reflectionGroups);
  const discussStages = importantReflectionGroups.map((reflectionGroup) => makeDiscussionStage(reflectionGroup.id, meetingId));

  if (!discussStages[0] || !discussPhase.stages[0]) return false;
  // mutative, replaces the placeholder with an actual stage
  discussStages[0].id = discussPhase.stages[0].id;
  discussPhase.stages = discussStages;

  await r.table('NewMeeting')
    .get(meetingId)
    .update({
      phases,
      updatedAt: now
    });
  const data = {teamId, meetingId};
  publish(TEAM, teamId, StartNewMeetingPayload, data, {operationId: subOptions.operationId});
  return true;
};

export default addDiscussionTopics;
