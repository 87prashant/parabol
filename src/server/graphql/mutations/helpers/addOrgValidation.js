import legitify from 'universal/validation/legitify'
import {fullName, orgName, requiredEmail, teamName} from 'universal/validation/templates'

export default function addOrgValidation () {
  return legitify({
    invitees: [
      {
        email: requiredEmail,
        fullName
      }
    ],
    newTeam: {
      name: teamName
    },
    orgName
  })
}
