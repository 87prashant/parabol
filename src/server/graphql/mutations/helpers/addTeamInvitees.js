import inviteTeamMembers from 'server/safeMutations/inviteTeamMembers';

const addTeamInvitees = async (invitees, teamId, viewerId) => {
  if (!invitees || invitees.length === 0) return {teamInviteNotifications: [], invitationIds: []};
  const {teamInviteNotifications, newInvitations} = await inviteTeamMembers(invitees, teamId, viewerId);
  const invitationIds = newInvitations.map(({id}) => id);
  return {teamInviteNotifications, invitationIds};
};

export default addTeamInvitees;
