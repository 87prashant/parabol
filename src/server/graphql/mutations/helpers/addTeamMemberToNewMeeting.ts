import createMeetingMembers from 'server/graphql/mutations/helpers/createMeetingMembers'
import {CHECKIN, UPDATES} from 'universal/utils/constants'
import getRethink from 'server/database/rethinkDriver'
import CheckInStage from 'server/database/types/CheckInStage'
import UpdatesStage from 'server/database/types/UpdatesStage'
import CheckInPhase from 'server/database/types/CheckInPhase'
import UpdatesPhase from 'server/database/types/UpdatesPhase'

/*
 * NewMeetings have a predefined set of stages, we need to add the new team member manually
 */

const setInPhase = (phase: CheckInPhase | UpdatesPhase, newStage: CheckInStage | UpdatesStage) => {
  const firstStage = phase.stages[0]
  newStage.isNavigable = firstStage.isNavigable
  newStage.isNavigableByFacilitator = firstStage.isNavigableByFacilitator
  phase.stages.push(newStage)
}

const addTeamMemberToNewMeeting = async (teamMember, teamId, dataLoader) => {
  const now = new Date()
  const r = getRethink()
  const team = await dataLoader.get('teams').load(teamId)
  const {meetingId} = team
  if (!meetingId) return false
  // make sure it's a new meeting
  const newMeeting = await r
    .table('NewMeeting')
    .get(meetingId)
    .default(null)
  if (!newMeeting) return false
  const {phases} = newMeeting
  const checkInPhase = phases.find((phase) => phase.phaseType === CHECKIN)
  const updatesPhase = phases.find((phase) => phase.phaseType === UPDATES)
  if (checkInPhase) {
    setInPhase(checkInPhase, new CheckInStage(teamMember.id))
  }
  if (updatesPhase) {
    setInPhase(updatesPhase, new UpdatesStage(teamMember.id))
  }
  const [meetingMember] = await createMeetingMembers(newMeeting, [teamMember], dataLoader)
  await r({
    meeting: r
      .table('NewMeeting')
      .get(meetingId)
      .update({
        phases,
        updatedAt: now
      }),
    member: r.table('MeetingMember').insert(meetingMember)
  })
  return true
}

export default addTeamMemberToNewMeeting
