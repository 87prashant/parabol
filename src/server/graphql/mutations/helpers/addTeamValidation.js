import legitify from 'universal/validation/legitify';
import {requiredId, fullName, requiredEmail, teamName} from 'universal/validation/templates';

export default function addTeamValidation() {
  return legitify({
    invitees: [{
      email: requiredEmail,
      fullName
    }],
    newTeam: {
      name: teamName,
      orgId: requiredId
    }
  });
}
