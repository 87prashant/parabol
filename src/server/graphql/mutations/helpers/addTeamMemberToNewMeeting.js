import {CHECKIN} from 'universal/utils/constants';
import getRethink from 'server/database/rethinkDriver';
import {makeCheckInStage} from 'server/graphql/mutations/helpers/makeCheckinStages';

/*
 * NewMeetings have a predefined set of stages, we need to add the new team member manually
 */
const addTeamMemberToNewMeeting = async (teamMember, teamId, dataLoader) => {
  const now = new Date();
  const r = getRethink();
  const team = await dataLoader.get('teams').load(teamId);
  const {meetingId} = team;
  if (!meetingId) return false;
  // make sure it's a new meeting
  const newMeeting = await r.table('NewMeeting').get(meetingId).default(null);
  if (!newMeeting) return false;
  const {phases} = newMeeting;
  const checkInPhase = phases.find((phase) => phase.phaseType === CHECKIN);
  if (!checkInPhase) return false;

  const {stages} = checkInPhase;
  const newStage = makeCheckInStage(teamMember, meetingId, false);
  stages.push(newStage);
  await r.table('NewMeeting')
    .get(meetingId)
    .update({
      phases,
      updatedAt: now
    });
  return true;
};

export default addTeamMemberToNewMeeting;
