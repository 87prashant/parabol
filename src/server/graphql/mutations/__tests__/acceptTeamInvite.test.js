import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockPubSub from 'server/__mocks__/MockPubSub';
import makeDataLoader from 'server/__tests__/setup/makeDataLoader';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import newInvitee from 'server/__tests__/utils/newInvitee';
import * as adjustUserCount from 'server/billing/helpers/adjustUserCount';
import getRethink from 'server/database/rethinkDriver';
import hashInviteTokenKey from 'server/graphql/mutations/helpers/inviteTeamMembers/hashInviteTokenKey';
import makeInviteToken from 'server/graphql/mutations/helpers/inviteTeamMembers/makeInviteToken';
import parseInviteToken from 'server/graphql/mutations/helpers/inviteTeamMembers/parseInviteToken';
import acceptTeamInvite from 'server/graphql/mutations/acceptTeamInvite';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {ADD_USER} from 'server/utils/serverConstants';
import * as tmsSignToken from 'server/utils/tmsSignToken';
import {TEAM_INVITE} from 'universal/utils/constants';

MockDate.set(__now);
console.error = jest.fn();

let auth0ManagementClient = null;

describe('acceptTeamInvite', () => {
  beforeEach(() => {
    auth0ManagementClient.users.updateAppMetadata.mockReset();
  });

  test.only('adds an invitee who was not previously in the org', async () => {
    // SETUP
    adjustUserCount.default = jest.fn();
    tmsSignToken.default = jest.fn(() => 'FAKEENCODEDJWT');
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    // const mockPubSub = new MockPubSub();
    const invitee = newInvitee('acceptTeamInvite1');
    const inviteToken = makeInviteToken();
    const {id} = parseInviteToken(inviteToken);
    const hashedToken = await hashInviteTokenKey(inviteToken);
    await mockDB.init()
      .newUser({name: 'inviteeGuy', tms: [], userOrgs: [], email: invitee.email})
      .newNotification(undefined, {type: TEAM_INVITE, email: invitee.email})
      .newInvitation({id, hashedToken, email: invitee.email})
      .newSoftTeamMember({email: invitee.email});

    const teamId = mockDB.context.team.id;
    const inviteeUser = mockDB.context.user;
    const authToken = mockAuthToken(inviteeUser);
    const dataLoader = makeDataLoader(authToken);
    const orgId = mockDB.context.organization.id;
    auth0ManagementClient.__initMock(mockDB.db);
    const tms = [teamId];
    // TEST
    await acceptTeamInvite.resolve(undefined, {inviteToken}, {authToken, dataLoader});

    // VERIFY
    const db = await fetchAndSerialize({
      user: r.table('User').get(inviteeUser.id),
      teamMember: r.table('TeamMember').get(`${inviteeUser.id}::${teamId}`),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds').coerceTo('array'),
      invitation: r.table('Invitation')
        .getAll(inviteeUser.email, {index: 'email'})
        .filter({teamId})
        .orderBy('tokenExpiration')
        .coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    // expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
    expect(adjustUserCount.default).toHaveBeenCalledWith(inviteeUser.id, orgId, ADD_USER);
    expect(auth0ManagementClient.users.updateAppMetadata).toHaveBeenCalledWith({id: inviteeUser.id}, {tms});
  });
  test('adds an invitee who was previously in the org', async () => {
    // SETUP
    adjustUserCount.default = jest.fn();
    tmsSignToken.default = jest.fn(() => 'FAKEENCODEDJWT');
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const mockPubSub = new MockPubSub();
    const inviteToken = makeInviteToken();
    const {id} = parseInviteToken(inviteToken);
    const hashedToken = await hashInviteTokenKey(inviteToken);
    await mockDB.init()
      .newTeam({orgId: mockDB.context.organization.id})
      .newNotification(undefined, {type: TEAM_INVITE, email: mockDB.db.user[1].email})
      .newInvitation({id, hashedToken, email: mockDB.db.user[1].email});
    const teamId = mockDB.context.team.id;
    const inviteeUser = mockDB.db.user[1];
    const authToken = mockAuthToken(inviteeUser);
    const dataLoader = makeDataLoader(authToken);
    const orgId = mockDB.context.organization.id;
    auth0ManagementClient.__initMock(mockDB.db);
    const tms = [mockDB.db.team[0].id, teamId];

    // TEST
    await acceptTeamInvite.resolve(undefined, {inviteToken}, {authToken, dataLoader});

    // VERIFY
    const db = await fetchAndSerialize({
      user: r.table('User').get(inviteeUser.id),
      teamMember: r.table('TeamMember').get(`${inviteeUser.id}::${teamId}`),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds').coerceTo('array'),
      invitation: r.table('Invitation')
        .getAll(inviteeUser.email, {index: 'email'})
        .filter({teamId})
        .orderBy('tokenExpiration')
        .coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
    expect(adjustUserCount.default).toHaveBeenCalledTimes(0);
    expect(auth0ManagementClient.users.updateAppMetadata).toHaveBeenCalledWith({id: inviteeUser.id}, {tms});
  });
});
