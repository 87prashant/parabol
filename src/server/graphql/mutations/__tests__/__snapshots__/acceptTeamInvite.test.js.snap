// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`acceptTeamInvite adds an invitee who was not previously in the org 1`] = `
Object {
  "invitation": Array [
    Object {
      "acceptedAt": 2017-01-01T08:00:00.000Z,
      "createdAt": 2017-01-01T07:00:00.000Z,
      "email": "<<constant>>",
      "hashedToken": "<<constant>>",
      "id": "<<constant>>",
      "inviteCount": 1,
      "invitedBy": "<<constant>>",
      "teamId": "<<constant>>",
      "tokenExpiration": 1970-01-01T00:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
  "notification": Array [],
  "teamMember": Object {
    "checkInOrder": 4,
    "email": "invitee@mockTeam+acceptTeamInvite1.co",
    "id": 3,
    "isCheckedIn": null,
    "isFacilitator": true,
    "isLead": false,
    "isNotRemoved": true,
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "inviteeGuy",
    "teamId": 1,
    "userId": 0,
  },
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "email": "invitee@mockTeam+acceptTeamInvite1.co",
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "inviteeGuy",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "inviteeGuy",
    "tms": Array [
      1,
    ],
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 2,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;
